src/ucp/proto/proto_select.c

```diff
@@ -25,6 +28,221 @@ typedef struct {
+    size_t midpoint;
+    char buf[64];
+
+    ucs_trace("candidate protocols for [%s]:",
+              ucs_memunits_range_str(start, end, buf, sizeof(buf)));
+    ucs_for_each_bit(curr.proto_id, proto_mask) {
+        ucs_trace("%24s %.0f+%.3f*X nsec",
+                  ucp_proto_id_field(curr.proto_id, name),
+                  proto_perf[curr.proto_id].c * UCS_NSEC_PER_SEC,
+                  proto_perf[curr.proto_id].m * UCS_NSEC_PER_SEC);
+    }
+
+    do {
+        ucs_assert(proto_mask != 0);
+
+        /* Find best protocol at the 'start' point */
+        best.result   = DBL_MAX;
+        best.proto_id = UCP_PROTO_ID_INVALID;
+        ucs_for_each_bit(curr.proto_id, proto_mask) {
+            curr.result = ucs_linear_func_apply(proto_perf[curr.proto_id],
+                                                start + 0.5);
+            if ((best.proto_id == UCP_PROTO_ID_INVALID) ||
+                (curr.result < best.result)) {
+                best = curr;
+            }
+        }
+
+        ucs_trace("  best protocol at %s is %s",
+                  ucs_memunits_to_str(start, buf, sizeof(buf)),
+                  ucp_proto_id_field(best.proto_id, name));
+
+        /* Find first (smallest) intersection point between the current best
+         * protocol and any other protocol. This would be the point where that
+         * other protocol becomes the best one.
+         */
+        midpoint    = end;
+        proto_mask &= ~UCS_BIT(best.proto_id);
+        ucs_for_each_bit(curr.proto_id, proto_mask) {
+            status = ucs_linear_func_intersect(proto_perf[curr.proto_id],
+                                               proto_perf[best.proto_id],
+                                               &x_intersect);
+            if ((status == UCS_OK) && (x_intersect > start)) {
+                /* We care only if the intersection is after 'start', since
+                 * otherwise best.proto_id is better than curr.proto_id at
+                 * 'end' as well as at 'start'.
+                 */
+                if (x_intersect < (double)SIZE_MAX) {
+                    midpoint = ucs_min((size_t)x_intersect, midpoint);
+                }
+                ucs_trace("   - intersects with %s at %.2f, midpoint is %s",
+                          ucp_proto_id_field(curr.proto_id, name), x_intersect,
+                          ucs_memunits_to_str(midpoint, buf, sizeof(buf)));
+            } else {
+                ucs_trace("   - intersects with %s out of range",
+                          ucp_proto_id_field(curr.proto_id, name));
+            }
+        }
+
+        status = ucp_proto_thresholds_append(thresh_list, midpoint,
+                                             best.proto_id);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        start = midpoint + 1;
+    } while (midpoint < end);
+    return UCS_OK;
+}
+
+/*
+ * Select a protocol for "msg_length", return last message length for the proto
```

Comment:
`"msg_lenght"` -> `'msg_length'` as in other cases, e.g.:
https://github.com/openucx/ucx/blob/4509003ab0d8557e80995d071677f26c44a7e4b0/src/ucp/proto/proto_select.c#L107